cloud/chatroom.js

```diff
@@ -40,3 +41,29 @@ function createChatRoom(maxOccupancy) {
 
     return chatRoom.save(null);
 }
+
+
+function getAvailableColorId(chatRoom) {
+    var aliasQuery = new Parse.Query("Alias");
+    aliasQuery.equalTo("chatRoomId", chatRoom.id);
+    aliasQuery.equalTo("active", true);
+
+    return aliasQuery.find().then(function(aliases){
+        var allColorIds = [];
```

Comment:
So `allColorIds` is pretty much just `[0,1,..,n-1]` where `n = chatRoom.get("maxOccupancy")` right?

Currently, you're building the list of `colorId`s in use and running a set intersection with `allColorIds`, but the `filter` is running in `O(n^2)` time. 

How about this:
```
var aliasQuery = ...
aliasQuery.equalTo(...); // Constraints
aliasQuery.ascending("colorId");
return aliasQuery.find().then(function(aliases) {
  // Step through the aliases that are now sorted by colorId.
  // We would expect the colorIds to be something like: [0,1,3,4] or something
  // so just check if the colorId matches the alias' index. if it doesn't, then
  // we know a number was skipped and therefore available for assignment
  for (var i = 0; i <= chatRoom.get("maxOccupancy") - 1; i++) {
    var colorId = aliases[i].get("colorId");
    if(colorId != i) {
      return Parse.Promise.as(colorId);
    }
  }
  // All colorIds have been assigned.
  return Parse.Promise.as(-1);
});